flips/component-interface.md

```diff
@@ -0,0 +1,267 @@
+  }
+
+  type ComponentManagerBuilder interface {
+    AddSubComponent(Component) ComponentManagerBuilder
+    AddChildRoutine(func(context.Context)) ComponentManagerBuilder
+    Build() *ComponentManager
+  }
+
+  func (c *ComponentManager) Start(ctx context.Context) error {
+    for _, component := range c.subComponents {
+      if err := component.Start(ctx); err != nil {
+        return err
+      }
+    }
+
+    for _, routine := range c.childRoutines {
+      c.childRoutinesStarted.Add(1)
+      c.childRoutinesFinished.Add(1)
+
+      go func(routine func(context.Context)) {
+        c.childRoutinesStarted.Done()
+        defer c.childRoutinesFinished.Done()
+
+        routine(ctx)
+      }(routine)
+    }
+  }
+
+  func (c *ComponentManager) Ready() <-chan struct{} {
+    ready := make(chan struct{})
+
+    go func() {
+      <-lifecycle.AllReady(c.subComponents...)
+      c.childRoutinesStarted.Wait()
+      close(ready)
+    }()
+
+    return ready
+  }
+
+  func (c *ComponentManager) Done() <-chan struct{} {
+    done := make(chan struct{})
+
+    go func() {
+      <-lifecycle.AllDone(c.subComponents...)
+      c.childRoutinesFinished.Wait()
+      close(done)
+    }()
+
+    return done
+  }
+  ```
+
+  Components that want to implement `Component` can use this `ComponentManager` to simplify implementation:
+
+  ```golang
+  type FooComponent struct {
+    *ComponentManager
+  }
+
+  func NewFooComponent(foo fooType, bar barType) *FooComponent {
+    f := &FooComponent{}
+
+    cmb := NewComponentManagerBuilder().
+      AddChildRoutine(f.childRoutine).
+      AddChildRoutine(f.childRoutineWithFooParameter(foo)).
+      AddSubComponent(NewBarComponent(bar))
+
+    f.ComponentManager = cmb.Build()
+
+    return f
+  }
+
+  func (f *FooComponent) childRoutine(ctx context.Context) {
+    for {
+      select {
+      case <-ctx.Done():
+        return
+      default:
+        // do work...
+      }
+    }
+  }
+
+  func (f *FooComponent) childRoutineWithFooParameter(foo fooType) func(context.Context) {
+    return func(ctx context.Context) {
+      for {
+        select {
+        case <-ctx.Done():
+          return
+        default:
+          // do work with foo...
+        }
+      }
+    }
+  }
+  ```
+
+* `ErrorAware` can be implemented with the help of an `ErrorManager` struct:
+
+  ```golang
+  // ErrorManager implements the ErrorAware interface, and provides a way for components
+  // to signal an irrecoverable error.
+  type ErrorManager struct {
+    errors chan error
+  }
+
+  func NewErrorManager() *ErrorManager {
+    return &ErrorManager{make(chan error)}
+  }
+
+  func (e *ErrorManager) Errors() <-chan error {
+    return e.errors
+  }
+
+  func (e *ErrorManager) ThrowError(err error) {
+    e.errors <- err
+    runtime.Goexit()
+  }
+  ```
+
+  Error handling for components can be implemented like so:
+
+  ```golang
+  type ErrorHandler func(ctx context.Context, errors <-chan error, restart func())
```

Comment:
Hey @AlexHentschel, thanks for the comment. Please see https://github.com/onflow/flow-go/pull/1275, which has a more up-to-date design for the error-handling part of this FLIP. I will actually be removing the error handling stuff from this FLIP, since it is already fully implemented in https://github.com/onflow/flow-go/pull/1275 and any further discussion can happen there.

Briefly, to answer your question, the idea is that if a component has experienced an irrecoverable error, it is not necessarily the case that it should *always* be restarted. For example, we may actually decide that the error cannot be fixed by a restart and should actually be propagated up the stack by throwing an irrecoverable error in the parent component as well. Or we may wish to do some additional cleanup / actions before restarting, such as firing an alert to pagerduty, logging the error, etc.

`ErrorHandler` gives the parent component the power to decide how to handle errors in the child component. The key here is that the *parent* is the one who should decide how to handle an irrecoverable error that occurs in the child component. An irrecoverable error is, by definition, an error that the child component doesn't know how to handle (if it did, the error would be recoverable), so all it can do is tell the parent component about the error and hope the parent knows how to handle it (whether that be via restarting, propagating the error further up the stack, etc).